sage_code/strong_uniform_bounds.py

```diff
@@ -0,0 +1,128 @@
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+    The authors can be reached at: barinder.s.banwait@gmail.com and
+    maarten@mderickx.nl.
+
+    ====================================================================
+
+"""
+
+from sage.all import Partitions, divisors, matrix, GF, lcm
+from itertools import product
+
+from .common_utils import get_weil_polys
+from .pil_integers import collapse_tuple
+
+def is_non_increasing(t):
+
+    for j in range(len(t)-1):
+        if t[j] < t[j+1]:
+            return False
+    return True
+
+
+def get_beta_mats_with_pow(F, pow=1):
+
+    frob_polys = get_weil_polys(F)
+    return [matrix.companion(f) ** pow for f in frob_polys]
+
+
+def remove_duplicates(list_of_e_tuples):
+
+    output = []
+
+    for t in list_of_e_tuples:
+        if is_non_increasing(t):
+            output.append(t)
+    return output
+
+def splitting_types(d):
+    """Implements step 1 of Algorithm 5.1, the possible splitting types
+
+    Args:
+        d (int): degree of number field
+    """
+
+    types = []
+
+    for r in range(1,d+1):
+        parts = list(Partitions(d, length=r))        
+        for a_part in parts:
+            divisors_of_partition = [divisors(t) for t in a_part]
+            possible_es = list(product(*divisors_of_partition))
+            possible_es = remove_duplicates(possible_es)
+            for possible_e_vec in possible_es:
+                f_vec = [a_part[j]/possible_e_vec[j] for j in range(r)]
+                type_dict = {}
+                type_dict['r'] = r
+                type_dict['es'] = possible_e_vec
+                type_dict['fs'] = tuple(f_vec)
+                types.append(type_dict)
+    return types
+
+
+def bound_from_split_type(split_type, eps, q):
+    """This implements Algorithm 5.1 from the paper, and returns the integers
+       B_{eps,q} and B^*_{eps,q}.
+
+    Args:
+        d (int): degree of number field
+        eps (tuple): isogeny signature
+        q (int): auxiliary rational prime
+    """
+    frob_poly_mats = [get_beta_mats_with_pow(GF(q**f), pow=my_e) for my_e,f in zip(split_type['es'],split_type['fs'])]
+    beta_mat_tuples = list(product(*frob_poly_mats))
+    collapsed_beta_mats = [
+        collapse_tuple(a_beta_tuple) for a_beta_tuple in beta_mat_tuples
+    ]
+    tr_eps = sum(eps)
+    q_to_tr_eps = q**tr_eps
+    running_lcm = 1
+    zero_detection_flag = False
+    for a_beta_mat in collapsed_beta_mats:
+        matrix_parent = a_beta_mat.parent()
+        alpha_to_eps_mat = matrix_parent(q_to_tr_eps)
+        pil_mat = alpha_to_eps_mat.tensor_product(a_beta_mat.parent()(1)) - (
+                  alpha_to_eps_mat.parent()(1)).tensor_product(a_beta_mat)
+        pil_int = pil_mat.det()
+        if pil_int == 0:
+            zero_detection_flag = True
+        else:
+            running_lcm = lcm(pil_int, running_lcm)
```

Comment:
This lcm might become big when q is big. Very often for q small this lcm is still computable. However for larger q this lcm might become so big that computing with integers of this size is really time consuming (or even making your computer run out of memory), and while profiling similar code from our previous article I noticed that without some optimization these integers became so big that even though computing lcm's is fast, it now became the most time consuming part of the algorithm. A good future optimisation (that doesn't need to be addressed in this MR) would be to have the ability to pass in a known multiplicative upperbound (obtained from smaller q) and gcd with this known multiplicative upperbound before doing the lcm. This keeps the size of these integers managable and leads to significant speedups.

Reply:
latest commit addresses these points